src/System.Buffers/src/System/Buffers/ArrayPoolBucket.cs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+
+namespace System.Buffers
+{
+    /// <summary>
+    /// Provides a thread-safe bucket containing buffers that can be Rented and Returned as part 
+    /// of a buffer pool; it should not be used independent of the pool.
+    /// </summary>
+    internal sealed class ArrayPoolBucket<T>
+    {
+        private volatile int _index;
+        private T[][] _data;
+        private SpinLock _lock;
+
+        /// <summary>
+        /// Creates the pool with numberOfBuffers arrays where each buffer is of bufferLength length.
+        /// </summary>
+        internal ArrayPoolBucket(int bufferLength, int numberOfBuffers)
+        {
+            _index = 0;
+            _lock = new SpinLock();
+            _data = new T[numberOfBuffers][];
+            for (int i = 0; i < numberOfBuffers; i++)
+            {
+                _data[i] = new T[bufferLength];
+            }
```

Comment:
Lots of good points in there...realistically this means that I can allocate each Bucket upfront and only lazy initialize each buffer. That will remove some of the interlocked operations in the Rent call, so this could have some perf improvements too. I'll make those changes now